api/api-environments-v3/src/main/java/com/thoughtworks/go/apiv3/environments/EnvironmentsControllerV3.java

```diff
@@ -0,0 +1,268 @@
+
+    @Autowired
+    public EnvironmentsControllerV3(ApiAuthenticationHelper apiAuthenticationHelper, EnvironmentConfigService environmentConfigService, EntityHashingService entityHashingService) {
+        super(ApiVersion.v3);
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+        this.environmentConfigService = environmentConfigService;
+        this.entityHashingService = entityHashingService;
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.Environments.BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerBasePath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+
+            before("", mimeType, apiAuthenticationHelper::checkAdminUserAnd403);
+            before("/*", mimeType, apiAuthenticationHelper::checkAdminUserAnd403);
+            get("", mimeType, this::index);
+            get(Routes.Environments.NAME, mimeType, this::show);
+            post("", mimeType, this::create);
+            put(Routes.Environments.NAME, mimeType, this::update);
+            patch(Routes.Environments.NAME, mimeType, this::partialUpdate);
+            delete(Routes.Environments.NAME, mimeType, this::remove);
+
+            exception(HttpException.class, this::httpException);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        Set<EnvironmentConfig> envConfigSet = environmentConfigService.getEnvironments();
+        List<EnvironmentConfig> envViewModelList = filterUnknownAndSortEnvConfigs(envConfigSet);
+        setEtagHeader(response, calculateEtag(envViewModelList));
+        return writerForTopLevelObject(request, response, outputWriter -> toJSON(outputWriter, envViewModelList));
+    }
+
+    public String show(Request request, Response response) throws IOException {
+        String environmentName = request.params("name");
+
+        EnvironmentConfig environmentConfig = fetchEntityFromConfig(environmentName);
+
+        setEtagHeader(environmentConfig, response);
+
+        return writerForTopLevelObject(request, response,
+                outputWriter -> EnvironmentRepresenter.toJSON(outputWriter, environmentConfig));
+    }
+
+    public String create(Request request, Response response) {
+        final BasicEnvironmentConfig environmentConfigToCreate = (BasicEnvironmentConfig) buildEntityFromRequestBody(request);
+        final HttpLocalizedOperationResult operationResult = new HttpLocalizedOperationResult();
+
+        if (!environmentConfigToCreate.getAllErrors().isEmpty()) {
+            operationResult.unprocessableEntity("Error parsing environment config from the request");
+            return handleCreateOrUpdateResponse(request, response, environmentConfigToCreate, operationResult);
+        }
+
+        haltIfEntityWithSameNameExists(environmentConfigToCreate);
```

Comment:
maybe, this check should be above of the request-body validation?

Reply:
I'm not in favor of this. The reason being `haltIfEntityWithSameNameExists` return the entity created from the request. And if the entity has some validation error - if may not be very clear.
This is my opinion. Let me know if u think otherwise.

Reply:
I am not a lot concerned with the order in which errors appear.
When I added that comment, I was looking at the following scenario:

As a user, I want to create a `prod` environment, so POST a JSON with pipelines and agents.
The response shows validation errors, which I fix and make another call, where the API request will again fail for saying "duplicate environment".

Now, the user can not proceed, as the environment that we wanted to create, already exists.

At this point, the initially fixed validation errors are of no use, as the error appeared next prevented user from creating an environment.

Reply:
Makes sense. Refactored.